modules/persistent-storage-csi-vsphere-rwx.adoc

```diff
@@ -0,0 +1,31 @@
+// Module included in the following assemblies:
+//
+// storage/container_storage_interface/persistent-storage-csi-vsphere.adoc
+//
+
+[id="persistent-storage-csi-vsphere-rwx_{context}"]
+= ReadWriteMany vSphere volume support
+
+If the underlying vSphere environment supports the vSAN file service, then vSphere Container Storage Interface (CSI) Driver Operator installed by
+{product-title} supports provisioning of ReadWriteMany (RWX) volumes. If vSAN file service is not configured, then ReadWriteOnce (RWO) is the only access mode available. If you do not have vSAN file service configured, and you request RWX, the volume fails to get created and an error is logged.
+
+For more information about configuring the vSAN file service in your environment, see https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.vsan.doc/GUID-82565B82-C911-42F7-85B1-E9EF973EE90C.html[vSAN File Service].
+
+You can request RWX volumes by making the following persistent volume claim (PVC):
+
+[source,yaml]
+----
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: myclaim
+spec:
+  resources:
+    requests:
+      storage: 1Gi
+  accessModes:
+     - ReadWriteMany
```

Comment:
Why isn't this managed by a [server side apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/), were this configuration is 'rejected' if the properly enabled configuration underneath the implementation isn't deliverable? 

If a pod fails to mount; because this isn't configured; isn't the VPC's configuration; vs what is produced (bonded) different ? How would a user know or be able to see this (by looking at the config) ? 
Would this be reconciled? If so would that reconciliation cause confusion with the user (as the requested value was changed by the controller/manager)?

Reply:
> Why isn't this managed by a [server side apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/), were this configuration is 'rejected' if the properly enabled configuration underneath the implementation isn't deliverable?

Server side is not the right answer for this. Server side apply is not a validation webhook. Its purpose is to let owners of objects specify their intent and apply them in server side (rather than sending patch objects).

We could in *theory* accomplish what you are saying with a webhook. It will be expensive and pros and cons has to be taken into account. Each PVC creation will result in validating capabilities of storage backend and hence will slow down PVC creation.  Traditionally - we have always handled this by not binding the PVC and adding appropriate event to PVC. We also don't want to special case just for `ReadWriteMany` volumes . What about a PVC request whose topology we can't satisfy? or a capacity request which of PVC which underlying storage can't satisfy? Should all such requests result in non-creation of PVCs?

Anyways - I think that, docs PR may not be right place for it. What you are proposing makes sense, but has to be handled by opening a KEP in `kubernetes/enhancements` project, if we are going to change mechanism of PVC creation itself.